src/react-render-perf/lesson-04/README.md

```diff
@@ -1,5 +1,102 @@
 [Home](../README.md)
 
-# 04 - Splitting Large Components
+# 04 - Splitting Up Large Components
 
-TODO
+Large components often end up re-rendering more stuff than is necessary when something
+changes in the component. If a component uses a lot of:
+- props
+- state
+- context usage
+- HOCs
+- hooks
+
+a change to any of these things will result in the entire component re-rendering.
+
+Splitting up a large component into a parent component and several smaller child
+components means that there will be fewer things (state, context usage, etc.) in
+the parent component that can trigger a re-render.  It also means that those things
+will be spread out across multiple children so when one of them changes only one of
+the children will re-render.
+
+Additionally, we can memoize some or all of the new child components if we want to
+prevent them from re-rendering in response props changes.
+
+## Example
+
+Consider `LargeComponent` below. Clicking on either of the buttons will cause
+the whole component to re-render.
+
+```tsx
+import {useState} from "react";
+
+const LargeComponent = () => {
+    const [foo, setFoo] = useState<number>(0);
+    const [bar, setBar] = useState<number>(0);
+
+    const foo = <>
+        <h1>foo = {foo}</h1>
+        <button onClick={incrementFoo}>increment foo</button>
+    </>;
+
+    const bar = <>
+        <h1>bar = {bar}</h1>
+        <button onClick={incrementBar}>increment bar</button>
+    </>;
+
+    return <>
+        <h1>LargeComponent</h1>
+        {foo}
+        {bar}
+    </>
+}
```

Comment:
Given the danger of people over-applying these lessons, well past the point where it's relevant for performance, what if we made one of these children quite expensive, so we would have more of a motivation for splitting it out? And then the "solution" would have only one split out, instead of the both of them.

Reply:
I'll add calls to `doExpensiveComputation()` or `<LargeSubTree />` or something like that to these components.